### ISSUE TITLE
Research directions of LLM

### ISSUE THREAD

**Issue by xp1632** [xp1632/Aalen_working_log#12]
After reading LowCoder and ViperGPT papers, I somehow don't know where I can contribute, the following is my thoughts before the discussion:

### Ori thoughts:

---
### LowCoder only generates one block instead of pipeline 

- LowCoder uses LLM to generate the recommended `Operator Name` instead of the whole pipeline 
- And the LowCoder generation of Only `Operator Name` is  around 70%, and the generation of `Operator Invocation Code` is even lower, around 40%
-  Based on this I can see why they landed their contribution on the `discoverability` of APIs in a visual programming tool
- because they cannot build a valid pipeline without manual manipulation of Operator blocks and with this low accuracy of generating `Operator Invocation Code`

-----

### Why LowCoder cannot generate `Pipeline`? 

**Training dataset issue from LowCoder Paper**
- I blame the low accuracy of the Lowcoder paper on their training set
- because their training set is too `fixed` for a llm
- they picked 79372 `one text query <-> one operator invocation code` pairs as training dataset
- trained one model from scratch and fine-tunned a T5 model
- then find while the model can only generate the code with 40% accuracy so let the model generate only operate name instead
- this  `one text query <-> one operator invocation code` pairs of training surely cannot let the model generate a pipeline 
- the model doesn't know how the operators will cooperate with each other, for example, the order of different operators for different tasks

---

### How to teach llm model the correct order of code components?

- As In ViperGPT paper, they provide the example code snippet in specification to let the model learn how to use their predefined APIs:
![image](https://github.com/user-attachments/assets/3c97fb7d-8491-4c4d-bd32-00266d3897e5)
![image](https://github.com/user-attachments/assets/ace53d71-9d8c-47de-9eb4-ce054a599dfa)

- so based on this, I thought, oh while our aim is to generate image processing pipeline, why don't we feed the specification of image processing libraries such as scikit, and let the model learn the correct order of the image processing operators?

- then I found out, during the pre-training of a llm model that can do code generation tasks, a large amount of python libraries include image processing libs such as scikit-image have already been trained

- what's the usage if we give the model the specs again ?
-  And not like concise several APIs in viperGPT, (find(), exist()), feeding a whole library specs will be very expensive and exceed the limit of context window
- so currently we only mark this problem here: https://github.com/xp1632/Aalen_working_log/issues/15

**Comment by xp1632** [comment]
### Current Thoughts 1: VLM before LLM, knowing `meta characteristics` of image input

**Image Segmentation Scenario**
- Tim tested the gpt4o model with grayscale, gradient noised image and a segment query 

- we find out the initial query gives the most general segment code snippet without considering the `characteristics` of the image itself
- for example, we have to tell the model in text that the image is `greyscale` after we find out the shape is wrong because it's set default as `RGB`
- then we have to tell the model the image to consider `gradient noising`, so it starts to consider gradient in the code 

---

**Add Vision Language Model for Image before language model**

- we notice that there's a lack of `understanding the basic characteristics of the image` for this gpt4o free version language model

- which is understandable for this model, because there's a high chance that this model is specialized in text area such as translation and summarization, not so good at image inference 

- so here we are, kind of like what ViperGPT did, let the `Vision language model` only deal with the understanding image part and let the language model only deal with code generation part

- we offer more `meta characteristics` we get from a Vision Language Model, and provide it to LLM so it can give more specific first response instead of general ones

![image](https://github.com/user-attachments/assets/b0dd8d02-74e2-4399-93b4-3d250270faec)


----

- However, this idea is similar to ViperGPT who use the different expertise for different VLS model and language model
- And doesn't current other multimodal llm already did this?

- Besides, the recent release of openAI showed great performance of image inference and query combination
- for example, turn a hand-drawing modularized pipeline into different part of feasible code 
- i think the free version of model we used today also somehow limit the performance

**Comment by xp1632** [comment]
### Current Thoughts 2: Chaldene Node pipeline JSON file

- let llm generate the JSON file that Chaldene can turn into a pipeline
- but again this is just format conversion problem 
- code --> block  is simple
- code snippet --> pipeline is hard just like the LowCoder tried to
